### PR TITLE
Phase 6: Fore-Foil Dedicated SRF Head (ID=6) — Split from Single-Foil

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

PR #2104 confirmed that giving the aft-foil (boundary ID=7) its own dedicated Surface Refinement Head improves **p_tan by -0.8%** (30.05 vs 30.29, 8-seed mean) — now merged into baseline. The aft-foil gets dedicated capacity because it operates in the wake with qualitatively different pressure distributions.

The **same logic applies to the fore-foil (boundary ID=6)**. Currently, the main `srf_head` serves both:
- ID=5 — single-foil surface nodes (simple, non-interactive flow)
- ID=6 — fore-foil tandem surface nodes (accelerated channel flow, modified lift from aft-foil presence)

The fore-foil in tandem sees aerodynamically distinct flow: gap acceleration increases local velocity, the aft foil's blockage modifies the upstream pressure distribution. A shared correction MLP with single-foil nodes cannot specialize to these effects. A **dedicated `fore_srf_head`** for ID=6 nodes should let the model learn richer fore-foil corrections, complementing the aft-foil SRF already merged.

**Key:** Zero-init on the last linear layer means worst-case this is identity — no regression risk. Single-foil samples have no ID=6 nodes so p_in is not affected.

**Prior work:** PR #2104 (merged) — dedicated aft-foil SRF. No prior experiment split fore-foil from single-foil.

## Instructions

You already have the SAF proxy from #2104 to identify aft-foil nodes (`saf_norm > 0.005` in tandem samples). Now extend the approach to split the remaining `srf_head` into single-foil (ID=5) and fore-foil (ID=6).

### Step 1 — Add config flags

```python
# In Config dataclass:
fore_foil_srf: bool = False
fore_foil_srf_hidden: int = 192
fore_foil_srf_layers: int = 3
```

### Step 2 — Init the fore-foil SRF head in `Transolver.__init__`

```python
if cfg.fore_foil_srf:
    self.fore_srf_head = SurfaceRefinementHead(
        n_hidden=cfg.n_hidden,
        hidden_size=cfg.fore_foil_srf_hidden,
        n_layers=cfg.fore_foil_srf_layers,
        out_dim=3,
    )
else:
    self.fore_srf_head = None
```

### Step 3 — Identify fore-foil nodes in `Transolver.forward`

Re-use your SAF proxy from #2104. You already compute `is_aft` (aft-foil = aft_mask). Now add:

```python
# Tandem indicator per sample (from gap feature, same as in #2104)
is_tandem_sample = (x[:, 0, 21].abs() > <your_threshold>)  # [B]
B, N = boundary_id_or_surface_mask.shape  # or however you get B, N

# Fore-foil: surface nodes in tandem samples that are NOT aft-foil
is_aft = <your existing aft-foil mask from #2104>  # [B, N]
is_fore = is_surface_mask & is_tandem_sample.unsqueeze(1).expand(-1, N) & ~is_aft  # [B, N]
```

### Step 4 — Narrow existing `srf_head` to single-foil only when --fore_foil_srf active

```python
if self.fore_srf_head is not None:
    # Narrow srf_head to single-foil only (non-tandem surface nodes)
    srf_apply_mask = is_surface_mask & ~is_tandem_sample.unsqueeze(1).expand(-1, N)
else:
    srf_apply_mask = is_surface_mask  # original: all surface nodes

# Apply srf_head using srf_apply_mask
if srf_apply_mask.any():
    srf_hidden = hidden[srf_apply_mask]
    srf_base = out[srf_apply_mask]
    srf_delta = self.srf_head(srf_hidden, srf_base)
    out[srf_apply_mask] += srf_delta
```

### Step 5 — Apply fore_srf_head to fore-foil (ID=6) nodes

```python
if self.fore_srf_head is not None and is_fore.any():
    fore_hidden = hidden[is_fore]
    fore_base = out[is_fore]
    fore_delta = self.fore_srf_head(fore_hidden, fore_base)
    out[is_fore] += fore_delta
```

### Step 6 — Run 6 experiments with `--wandb_group phase6/fore-foil-srf`

**Important:** The new baseline includes `--aft_foil_srf`. All runs must include it.

| GPU | Config | wandb_name | Seed |
|-----|--------|-----------|------|
| 0 | Baseline (aft_srf only, control) | fern/fore-srf-base-s42 | 42 |
| 1 | Baseline (aft_srf only, control) | fern/fore-srf-base-s43 | 43 |
| 2 | `--fore_foil_srf` (192/3L) | fern/fore-srf-s42 | 42 |
| 3 | `--fore_foil_srf` (192/3L) | fern/fore-srf-s43 | 43 |
| 4 | `--fore_foil_srf --fore_foil_srf_hidden 256 --fore_foil_srf_layers 4` | fern/fore-srf-lg-s42 | 42 |
| 5 | `--fore_foil_srf --fore_foil_srf_hidden 256 --fore_foil_srf_layers 4` | fern/fore-srf-lg-s43 | 43 |

Full base command (all runs include **both** `--aft_foil_srf` and optionally `--fore_foil_srf`):
```bash
cd cfd_tandemfoil && nohup env PYTHONUNBUFFERED=1 CUDA_VISIBLE_DEVICES=<gpu> python train.py \
  --agent fern --wandb_name "fern/<name>" --wandb_group phase6/fore-foil-srf \
  --seed <seed> \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --disable_pcgrad --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf \
  [--fore_foil_srf] [--fore_foil_srf_hidden 256 --fore_foil_srf_layers 4] \
  > logs/<name>.log 2>&1 &
```

**Do NOT override SENPAI_MAX_EPOCHS or SENPAI_TIMEOUT_MINUTES.**

## What to report

- Surface MAE for each config: p_in, p_oodc, p_tan, p_re (and W&B run IDs)
- Does `--fore_foil_srf` improve p_tan? By how much vs. control baseline?
- Does larger capacity (256/4L) help vs default (192/3L)?
- Is p_in stable? (Expected: yes — fore-foil head only fires on ID=6 tandem nodes)
- How did you identify fore-foil nodes? (Expected: `is_surface & is_tandem & ~is_aft_foil`)
- Node count sanity check: how many fore-foil vs aft-foil vs single-foil nodes per batch?

## Baseline

Current best (PR #2104 merged — +aft_foil_srf, 8-seed mean, seeds 42-49):

| Metric | 8-seed mean (aft_srf) | Target to beat |
|--------|----------------------|----------------|
| p_in | 13.19 ± 0.33 | < 13.19 |
| p_oodc | 7.92 ± 0.17 | < 7.92 |
| p_tan | **30.05 ± 0.36** | **< 30.05** |
| p_re | 6.45 ± 0.07 | < 6.45 |

For merge decisions, compare your 2-seed avg against the 8-seed mean (p_tan < 30.05 is primary target).

Reproduce baseline:
```bash
cd cfd_tandemfoil && python train.py --agent <name> --wandb_name "baseline" \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --disable_pcgrad --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf
```

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT fern:

## Results

### Experiment Summary

| Config | Seed | p_in | p_oodc | p_tan | p_re | val/loss | Epochs | W&B Run |
|--------|------|------|--------|-------|------|----------|--------|---------|
| Baseline (aft_srf only) | 42 | 13.4 | 7.6 | 29.6 | 6.4 | 0.3856 | 155 | sqfe4eih |
| Baseline (aft_srf only) | 43 | 13.2 | 7.8 | 29.3 | 6.4 | 0.3860 | 155 | 69va7w4w |
| **Baseline avg** | — | **13.3** | **7.7** | **29.45** | **6.4** | **0.3858** | — | — |
| +fore_srf (192/3L) | 42 | 13.2 | 7.7 | 32.1 | 6.4 | 0.3947 | 153 | oiygmwit |
| +fore_srf (192/3L) | 43 | 13.2 | 7.8 | 32.2 | 6.5 | 0.3966 | 152 | o3ziow82 |
| **fore_srf avg** | — | **13.2** | **7.75** | **32.15** | **6.45** | **0.3957** | — | — |
| +fore_srf_lg (256/4L) | 42 | 13.8 | 8.0 | 32.4 | 6.7 | 0.3971 | 151 | qkp9400n |
| +fore_srf_lg (256/4L) | 43 | 13.1 | 8.0 | 33.1 | 6.5 | 0.3999 | 152 | w62a6akm |
| **fore_srf_lg avg** | — | **13.45** | **8.0** | **32.75** | **6.6** | **0.3985** | — | — |

**PR baseline (8-seed mean):** p_in=13.19, p_oodc=7.92, p_tan=30.05, p_re=6.45

### Key Findings

**This is a clear negative result.** The dedicated fore-foil SRF head significantly degrades p_tan:

- **fore_srf 192/3L:** p_tan = 32.15 vs baseline 29.45 → **+9.2% worse**
- **fore_srf_lg 256/4L:** p_tan = 32.75 vs baseline 29.45 → **+11.2% worse**
- The larger capacity (256/4L) is even worse than default (192/3L)

**p_in is stable** for the 192/3L variant (13.2 vs 13.3), as expected since fore-foil nodes don't exist in single-foil samples. The 256/4L variant shows slight p_in regression (13.45), likely from optimizer competition.

**p_oodc slightly degraded** across both variants (7.75–8.0 vs baseline 7.7).

### Analysis: Why the fore-foil head hurts

The key issue is **splitting the srf_head**. When `--fore_foil_srf` is active, the existing `srf_head` is narrowed to single-foil surface nodes only — it no longer sees tandem fore-foil nodes during training. This means:

1. **The srf_head loses transfer learning from tandem data.** The shared head previously learned surface corrections from both single-foil and tandem fore-foil nodes, benefiting from the larger effective training set.
2. **The dedicated fore_srf_head sees far fewer nodes** — only ~200 fore-foil nodes per tandem sample, vs ~400+ surface nodes per sample for the shared head. With the tandem split being only ~40% of training data, the fore-foil head gets limited signal.
3. **The zero-init safety net means worst case is identity, but the head still participates in the optimizer** and gradient dynamics. The narrowed srf_head + dedicated fore_srf_head combination is strictly worse than the shared srf_head.

This is the opposite of the aft-foil SRF result: the aft-foil head works because boundary ID=7 has qualitatively different flow (wake region), so specialization helps. The fore-foil (ID=6) apparently doesn't need the same specialization — the shared srf_head already handles it well.

### Fore-foil node identification

Fore-foil mask: `is_surface & is_tandem & ~is_aft_foil_mask` where:
- `is_tandem = (gap_feature.abs() > 0.01)`
- `is_aft_foil_mask = is_surface & (saf_norm > 0.005) & is_tandem`

This correctly identifies all boundary ID=6 nodes in tandem samples.

### Experiment Details

- All runs: 180 min timeout, ~152-155 epochs
- Peak VRAM: 38.4 GB (identical across configs — fore_srf head is tiny vs model)
- Base command included `--aft_foil_srf` for all runs (new baseline)
- wandb_group: `phase6/fore-foil-srf`

### Suggested Follow-ups

1. **Don't split the srf_head.** Instead of narrowing srf_head to single-foil only, try adding the fore_srf_head as a *second additive correction* on top of the shared srf_head (i.e., stacked refinement). This preserves transfer learning while giving extra capacity.
2. **Fore-foil-specific loss weighting** instead of a dedicated head — upweight fore-foil surface nodes in the existing loss computation to give the model more incentive to get them right.
3. **Investigate whether fore-foil pressure is actually the bottleneck** for p_tan. It could be that aft-foil pressure dominates the tandem MAE and fore-foil is already well-predicted.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2117 because DEAD END — dedicated fore-foil SRF head (splitting approach) degrades p_tan by +9-11%.

Root cause is clear: by narrowing the shared `srf_head` to single-foil nodes only when `--fore_foil_srf` is active, the head loses the transfer learning signal from tandem data. The fore-foil correction patterns are similar enough to single-foil that the shared head was already handling them well — specialization without splitting would be the correct formulation.

Key contrast with the successful aft-foil SRF (#2104): the aft-foil head was ADDED as an extra correction on top of the shared srf_head (not replacing it), and the aft-foil flow is qualitatively different (wake region, reversed pressure gradient). The fore-foil cannot benefit from the same splitting approach.

Excellent analysis from fern — the stacked refinement suggestion (add fore_srf_head additively on top of shared srf_head, without narrowing it) is the correct next step and will be assigned as a new experiment.

Thank you for thorough control runs and clear root-cause diagnosis..
